### PR TITLE
runelite-api: Add missing documentation

### LINKED ISSUE #500: Add/improve API documentation
Currently, the documentation in API is almost absent, so it is sometimes hard to understand what is going on (also, commented code with JavaDocs looks very nice).

[Current runelite-api documentation](http://static.runelite.net/api/runelite-api/)
[Current runelite-client documentation](http://static.runelite.net/api/runelite-client/)

Except net.runelite.client.plugins, all fields in these APIs should be documented. We probably need to start creating PRs documenting class by class (or smaller packages) to not create giant PR, so any contributions to solving at least parts of this issue are welcome. 

Also, it is very good issue for first contribution to the project/testing the waters.

### PR DESCRIPTION
**Pull request by sainttx**
Adds much needed javadoc documentation comments to methods in the API and resolves #500.

Although the diff is very large and might need to be split up, I thought it would be a good start to have everything centralized for review in one place. If there are any additions, changes, or feedbacks about the proposed documentation please leave comments.

Current incomplete classes are:
- Script
- TextureProvider
- PacketBuffer
- OverlayIndex

### CODE REVIEW

**Comment by sainttx** [comment]
@deathbeam fixed

File: runelite-api/src/main/java/net/runelite/api/GraphicID.java [hunk 1]
```diff
@@ -26,13 +26,18 @@
 
 public class GraphicID
 {
+	public static final int TELEPORT_SPELL = 111;
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/GraphicID.java]: This should be probably split in different commit.

File: runelite-api/src/main/java/net/runelite/api/events/MenuOptionClicked.java [hunk 2]
```diff
@@ -25,19 +25,60 @@
...
+	/**
+	 * The action performed.
+	 */
 	private MenuAction menuAction;
+	/**
+	 * The ID of the object, actor, or item that the interaction targets.
+	 */
 	private int id;
+	/**
+	 * The ID of the widget that.
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/MenuOptionClicked.java]: The id of the widget what?

File: runelite-api/src/main/java/net/runelite/api/events/NpcDespawned.java [hunk 3]
```diff
@@ -25,16 +25,21 @@
...
 public class NpcDespawned implements ActorDespawned
 {
+	/**
+	 * The despawned NPC.
+	 */
 	private final NPC npc;
 
 	@Override
-	public Actor getActor()
+	public NPC getActor()
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/NpcDespawned.java]: Why this change?

File: runelite-api/src/main/java/net/runelite/api/events/NpcSpawned.java [hunk 4]
```diff
@@ -25,16 +25,21 @@
...
 public class NpcSpawned implements ActorSpawned
 {
+	/**
+	 * The spawned NPC.
+	 */
 	private final NPC npc;
 
 	@Override
-	public Actor getActor()
+	public NPC getActor()
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/NpcSpawned.java]: Why this change?

File: runelite-api/src/main/java/net/runelite/api/events/PlayerDespawned.java [hunk 5]
```diff
@@ -25,16 +25,21 @@
...
 public class PlayerDespawned implements ActorDespawned
 {
+	/**
+	 * The despawned player.
+	 */
 	private final Player player;
 
 	@Override
-	public Actor getActor()
+	public Player getActor()
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/PlayerDespawned.java]: Why this change?

File: runelite-api/src/main/java/net/runelite/api/events/PlayerSpawned.java [hunk 6]
```diff
@@ -25,16 +25,21 @@
...
 public class PlayerSpawned implements ActorSpawned
 {
+	/**
+	 * The spawned player.
+	 */
 	private final Player player;
 
 	@Override
-	public Actor getActor()
+	public Player getActor()
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/PlayerSpawned.java]: Why this change

File: runelite-api/src/main/java/net/runelite/api/events/WidgetPositioned.java [hunk 7]
```diff
@@ -26,6 +26,9 @@
 
 import lombok.Value;
 
+/**
+ *
+ */
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/events/WidgetPositioned.java]: Nothing?

**Comment by sainttx** [comment]
@deathbeam The javadoc in MenuOptionClicked and WidgetPositioned has been fixed.

My reasoning for changing the method return value in the `NpcSpawned`, `NpcDespawned`, `PlayerSpawned`, `PlayerDespawned` events is to simplify use cases when using those events. For example, if we subscribe to the `PlayerSpawned` event we  should be able to safely assume that the `Actor` spawning is a `Player` entity (based on the event name). If we had a call like `Player spawned = (Player) event.getActor()`, we can remove the redundant cast by returning `Player` in the event since it makes sense. With the redundancy removed we have `Player spawned = event.getActor()`. If we want to subscribe to all `ActorSpawned` and `ActorDespawned` we can still do `Actor actor = event.getActor()` so it makes sense logically to narrow the scope for the subclasses.

File: runelite-api/src/main/java/net/runelite/api/events/MenuOptionClicked.java [hunk 8]
```diff
@@ -25,19 +25,62 @@
...
+ * <p>
+ * This event does not only trigger when clicking an option in a
+ * right-clicked menu. The event will trigger for any left click
+ * action performed (ie. clicking an item, walking to a tile, etc) as
+ * well as any right-click menu option.
+ * <p>
+ * By default, when there is no action performed when left-clicking,
+ * it seems that this event still triggers with the "Cancel" action.
+ */
+@ToString
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/events/MenuOptionClicked.java]: `@Data` implies `@ToString`

**Comment by sainttx** [comment]
@Adam- yeah took quite a bit of work and digging through code to understand certain methods but it was pretty fun. I've reverted the changes requested by @deathbeam for now and will submit them in a future PR.

One question I had actually is about the [KitType](https://github.com/runelite/runelite/blob/master/runelite-api/src/main/java/net/runelite/api/kit/KitType.java) and [EquipmentInventorySlot](https://github.com/runelite/runelite/blob/master/runelite-api/src/main/java/net/runelite/api/EquipmentInventorySlot.java) classes. They both seem to do the same thing (ie. same indexes for equipment slots) and KitType is even used in `PlayerComposition.getEquipmentId(KitType)`, is having both classes necessary?

**Comment by Adam-** [comment]
KitType is the index in the player composition equipments while EquipmentInventorySlot is the index in the inventory for the equipment. I think they are two different things although it is interesting the indexes are seemingly the same.

**Comment by sainttx** [comment]
@Adam- Cheers, have updated the documentation for `EquipmentInventorySlot` and `KitType` and merged the latest updates to the API. Let me know if there is anything else that needs to be done.

**Comment by Adam-** [comment]
Thanks!
